File: pravega/src/test/scala/docs/scaladsl/PravegaReadWriteDocs.scala
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl

import akka.actor.ActorSystem
import akka.stream.alpakka.pravega.{
  PravegaEvent,
  ReaderSettingsBuilder,
  TableWriterSettingsBuilder,
  WriterSettingsBuilder
}
import akka.stream.scaladsl.{Sink, Source}
import io.pravega.client.ClientConfig
import io.pravega.client.stream.Serializer
import io.pravega.client.stream.impl.UTF8StringSerializer

import java.nio.ByteBuffer
import akka.stream.alpakka.pravega.TableReaderSettingsBuilder
import akka.stream.alpakka.pravega.scaladsl.PravegaTable
import akka.stream.alpakka.pravega.scaladsl.Pravega
import scala.util.Using
import io.pravega.client.tables.TableKey

class PravegaReadWriteDocs {

  implicit val system = ActorSystem("PravegaDocs")

  val serializer = new UTF8StringSerializer

  implicit def personSerialiser: Serializer[Person] = ???

  implicit val intSerializer = new Serializer[Int] {
    override def serialize(value: Int): ByteBuffer = {
      val buff = ByteBuffer.allocate(4).putInt(value)
      buff.position(0)
      buff
    }

    override def deserialize(serializedValue: ByteBuffer): Int =
      serializedValue.getInt
  }

  val readerSettings = ReaderSettingsBuilder(system)
    .withSerializer(serializer)

  val writerSettings = WriterSettingsBuilder(system)
    .withSerializer(serializer)

  val writerSettingsWithRoutingKey = WriterSettingsBuilder(system)
    .withKeyExtractor((str: String) => str.take(1))
    .withSerializer(serializer)

  // #writing
  Source(1 to 100)
    .map(i => s"event_$i")
    .runWith(Pravega.sink("an_existing_scope", "an_existing_streamName", writerSettings))

  Source(1 to 100)
    .map { i =>
      val routingKey = i % 10
      s"${routingKey}_event_$i"
    }
    .runWith(Pravega.sink("an_existing_scope", "an_existing_streamName", writerSettingsWithRoutingKey))

  // #writing

  def processMessage(message: String): Unit = ???

  // #reader-group

  Using(Pravega.readerGroupManager("an_existing_scope", readerSettings.clientConfig)) { readerGroupManager =>
    readerGroupManager.createReaderGroup("myGroup", "stream1", "stream2")
  }
  // #reader-group
    .foreach { readerGroup =>
      // #reading

      Pravega
        .source(readerGroup, readerSettings)
        .to(Sink.foreach { event: PravegaEvent[String] =>
          val message: String = event.message
          processMessage(message)
        })
        .run()

      // #reading

    }

  implicit val tablewriterSettings = TableWriterSettingsBuilder[Int, Person]()
    .withKeyExtractor(id => new TableKey(intSerializer.serialize(id)))
    .build()

  // #table-writing

  // Write through a flow
  Source(1 to 10)
    .map(id => (id, Person(id, s"name_$id")))
    .via(PravegaTable.writeFlow("an_existing_scope", "an_existing_tablename", tablewriterSettings))
    .runWith(Sink.ignore)

  // Write in a sink
  Source(1 to 10)
    .map(id => (id, Person(id, s"name_$id")))
    .runWith(PravegaTable.sink("an_existing_scope", "an_existing_tablename", tablewriterSettings))

  // #table-writing

  val clientConfig = ClientConfig.builder().build()

  val tableSettings = TableReaderSettingsBuilder[Int, Person]()
    .withKeyExtractor(id => new TableKey(intSerializer.serialize(id)))
    .build()

  // #table-reading

  val readingDone = PravegaTable
    .source("an_existing_scope", "an_existing_tablename", tableSettings)
    .to(Sink.foreach(println))
    .run()

  // #table-reading

}

File: pravega/src/test/java/docs/javadsl/PravegaReadWriteDocs.java
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://akka.io>
 */

package docs.javadsl;

import akka.Done;
import akka.japi.Pair;
import akka.stream.alpakka.pravega.*;
import akka.stream.alpakka.pravega.PravegaReaderGroupManager;
import akka.stream.alpakka.pravega.javadsl.Pravega;
import akka.stream.alpakka.pravega.javadsl.PravegaTable;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import io.pravega.client.stream.ReaderGroup;
import io.pravega.client.stream.Serializer;
import io.pravega.client.stream.impl.JavaSerializer;
import io.pravega.client.stream.impl.UTF8StringSerializer;
import io.pravega.client.tables.TableKey;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletionStage;

public class PravegaReadWriteDocs extends PravegaAkkaTestCaseSupport {

  public static void docs() {
    WriterSettings<String> writerSettings =
        WriterSettingsBuilder.<String>create(system).withSerializer(new JavaSerializer<>());
    WriterSettings<String> writerSettingsWithRoutingKey =
        WriterSettingsBuilder.<String>create(system)
            .withKeyExtractor((String str) -> str.substring(0, 2))
            .withSerializer(new JavaSerializer<>());
    ReaderSettings<String> readerSettings =
        ReaderSettingsBuilder.create(system).withSerializer(new JavaSerializer<>());

    // #writing
    Sink<String, CompletionStage<Done>> sinkWithRouting =
        Pravega.sink("an_existing_scope", "an_existing_scope", writerSettings);

    CompletionStage<Done> doneWithRouting =
        Source.from(Arrays.asList("One", "Two", "Three")).runWith(sinkWithRouting, system);

    // #writing

    // #reader-group

    ReaderGroup readerGroup;
    try (PravegaReaderGroupManager readerGroupManager =
        Pravega.readerGroup("an_existing_scope", readerSettings.clientConfig())) {
      readerGroup = readerGroupManager.createReaderGroup("my_group", "streamName");
    }
    // #reader-group

    // #reading
    CompletionStage<Done> fut =
        Pravega.<String>source(readerGroup, readerSettings)
            .to(Sink.foreach(e -> processMessage(e.message())))
            .run(system);
    // #reading

    UTF8StringSerializer serializer = new UTF8StringSerializer();

    Serializer<Integer> intSerializer =
        new Serializer<Integer>() {
          public ByteBuffer serialize(Integer value) {
            ByteBuffer buff = ByteBuffer.allocate(4).putInt(value);
            buff.position(0);
            return buff;
          }

          public Integer deserialize(ByteBuffer serializedValue) {

            return serializedValue.getInt();
          }
        };

    TableWriterSettings<Integer, String> tablewriterSettings =
        TableWriterSettingsBuilder.<Integer, String>create(system, intSerializer, serializer)
            .withKeyExtractor(id -> new TableKey(intSerializer.serialize(id)))
            .build();

    // #table-writing
    final List<Pair<Integer, String>> events =
        Arrays.asList(
            new Pair<Integer, String>(1, "One"),
            new Pair<Integer, String>(2, "Two"),
            new Pair<Integer, String>(3, "Three"),
            new Pair<Integer, String>(4, "Four"));

    Sink<Pair<Integer, String>, CompletionStage<Done>> sink =
        PravegaTable.sink("an_existing_scope", "an_existing_tableName", tablewriterSettings);

    CompletionStage<Done> done = Source.from(events).toMat(sink, Keep.right()).run(system);

    // #table-writing

    TableReaderSettings<Integer, String> tableReaderSettings =
        TableReaderSettingsBuilder.<Integer, String>create(system, intSerializer, serializer)
            .withKeyExtractor(id -> new TableKey(intSerializer.serialize(id)))
            .build();

    // #table-reading

    final CompletionStage<Done> pair =
        PravegaTable.source("an_existing_scope", "an_existing_tableName", tableReaderSettings)
            .to(Sink.foreach((TableEntry<String> kvp) -> processKVP(kvp)))
            .run(system);
    // #table-reading

  }

  private static void processKVP(TableEntry<String> kvp) {
    LOGGER.info(kvp.toString());
  }

  private static void processMessage(String message) {
    LOGGER.info(message);
  }
}
